// @pulumi/kubernetes/rbac/v1/clusterRole.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * ClusterRole is a cluster level, logical grouping of PolicyRules that can be referenced as a unit by a RoleBinding or ClusterRoleBinding.
 */
class ClusterRole extends pulumi.CustomResource {
    /**
     * Create a ClusterRole resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let inputs = {};
        if (!(opts && opts.id)) {
            inputs["aggregationRule"] = args ? args.aggregationRule : undefined;
            inputs["apiVersion"] = "rbac.authorization.k8s.io/v1";
            inputs["kind"] = "ClusterRole";
            inputs["metadata"] = args ? args.metadata : undefined;
            inputs["rules"] = args ? args.rules : undefined;
        }
        else {
            inputs["aggregationRule"] = undefined /*out*/;
            inputs["apiVersion"] = undefined /*out*/;
            inputs["kind"] = undefined /*out*/;
            inputs["metadata"] = undefined /*out*/;
            inputs["rules"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        const aliasOpts = { aliases: [{ type: "kubernetes:rbac.authorization.k8s.io/v1alpha1:ClusterRole" }, { type: "kubernetes:rbac.authorization.k8s.io/v1beta1:ClusterRole" }] };
        opts = opts ? pulumi.mergeOptions(opts, aliasOpts) : aliasOpts;
        super(ClusterRole.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ClusterRole resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new ClusterRole(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ClusterRole.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ClusterRole.__pulumiType;
    }
}
exports.ClusterRole = ClusterRole;
/** @internal */
ClusterRole.__pulumiType = 'kubernetes:rbac.authorization.k8s.io/v1:ClusterRole';
