// @pulumi/kubernetes/rbac/v1/clusterRoleBinding.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * ClusterRoleBinding references a ClusterRole, but not contain it.  It can reference a ClusterRole in the global namespace, and adds who information via Subject.
 */
class ClusterRoleBinding extends pulumi.CustomResource {
    /**
     * Create a ClusterRoleBinding resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let inputs = {};
        if (!(opts && opts.id)) {
            if ((!args || args.roleRef === undefined) && !(opts && opts.urn)) {
                throw new Error("Missing required property 'roleRef'");
            }
            inputs["apiVersion"] = "rbac.authorization.k8s.io/v1";
            inputs["kind"] = "ClusterRoleBinding";
            inputs["metadata"] = args ? args.metadata : undefined;
            inputs["roleRef"] = args ? args.roleRef : undefined;
            inputs["subjects"] = args ? args.subjects : undefined;
        }
        else {
            inputs["apiVersion"] = undefined /*out*/;
            inputs["kind"] = undefined /*out*/;
            inputs["metadata"] = undefined /*out*/;
            inputs["roleRef"] = undefined /*out*/;
            inputs["subjects"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        const aliasOpts = { aliases: [{ type: "kubernetes:rbac.authorization.k8s.io/v1alpha1:ClusterRoleBinding" }, { type: "kubernetes:rbac.authorization.k8s.io/v1beta1:ClusterRoleBinding" }] };
        opts = opts ? pulumi.mergeOptions(opts, aliasOpts) : aliasOpts;
        super(ClusterRoleBinding.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing ClusterRoleBinding resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new ClusterRoleBinding(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ClusterRoleBinding.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ClusterRoleBinding.__pulumiType;
    }
}
exports.ClusterRoleBinding = ClusterRoleBinding;
/** @internal */
ClusterRoleBinding.__pulumiType = 'kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding';
