// @pulumi/kubernetes/core/v1/podTemplate.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * PodTemplate describes a template for creating copies of a predefined pod.
 */
class PodTemplate extends pulumi.CustomResource {
    /**
     * Create a PodTemplate resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let inputs = {};
        if (!(opts && opts.id)) {
            inputs["apiVersion"] = "v1";
            inputs["kind"] = "PodTemplate";
            inputs["metadata"] = args ? args.metadata : undefined;
            inputs["template"] = args ? args.template : undefined;
        }
        else {
            inputs["apiVersion"] = undefined /*out*/;
            inputs["kind"] = undefined /*out*/;
            inputs["metadata"] = undefined /*out*/;
            inputs["template"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(PodTemplate.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing PodTemplate resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new PodTemplate(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of PodTemplate.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === PodTemplate.__pulumiType;
    }
}
exports.PodTemplate = PodTemplate;
/** @internal */
PodTemplate.__pulumiType = 'kubernetes:core/v1:PodTemplate';
