// @pulumi/kubernetes/provider.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
/**
 * The provider type for the kubernetes package.
 */
class Provider extends pulumi.ProviderResource {
    /**
     * Create a Provider resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        var _a, _b, _c;
        let inputs = {};
        {
            inputs["cluster"] = args ? args.cluster : undefined;
            inputs["context"] = args ? args.context : undefined;
            inputs["enableDryRun"] = pulumi.output((_a = ((args ? args.enableDryRun : undefined) || utilities.getEnvBoolean("PULUMI_K8S_ENABLE_DRY_RUN")), (_a !== null && _a !== void 0 ? _a : utilities.getEnvBoolean("PULUMI_K8S_ENABLE_DRY_RUN")))).apply(JSON.stringify);
            inputs["kubeconfig"] = (_b = ((args ? args.kubeconfig : undefined) || utilities.getEnv("KUBECONFIG")), (_b !== null && _b !== void 0 ? _b : utilities.getEnv("KUBECONFIG")));
            inputs["namespace"] = args ? args.namespace : undefined;
            inputs["renderYamlToDirectory"] = args ? args.renderYamlToDirectory : undefined;
            inputs["suppressDeprecationWarnings"] = pulumi.output((_c = ((args ? args.suppressDeprecationWarnings : undefined) || utilities.getEnvBoolean("PULUMI_K8S_SUPPRESS_DEPRECATION_WARNINGS")), (_c !== null && _c !== void 0 ? _c : utilities.getEnvBoolean("PULUMI_K8S_SUPPRESS_DEPRECATION_WARNINGS")))).apply(JSON.stringify);
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(Provider.__pulumiType, name, inputs, opts);
    }
    /**
     * Returns true if the given object is an instance of Provider.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Provider.__pulumiType;
    }
}
exports.Provider = Provider;
/** @internal */
Provider.__pulumiType = 'kubernetes';
