// @pulumi/kubernetes/policy/v1beta1/podDisruptionBudget.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * PodDisruptionBudget is an object to define the max disruption that can be caused to a collection of pods
 */
class PodDisruptionBudget extends pulumi.CustomResource {
    /**
     * Create a PodDisruptionBudget resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let inputs = {};
        if (!(opts && opts.id)) {
            inputs["apiVersion"] = "policy/v1beta1";
            inputs["kind"] = "PodDisruptionBudget";
            inputs["metadata"] = args ? args.metadata : undefined;
            inputs["spec"] = args ? args.spec : undefined;
            inputs["status"] = undefined /*out*/;
        }
        else {
            inputs["apiVersion"] = undefined /*out*/;
            inputs["kind"] = undefined /*out*/;
            inputs["metadata"] = undefined /*out*/;
            inputs["spec"] = undefined /*out*/;
            inputs["status"] = undefined /*out*/;
        }
        if (!opts) {
            opts = {};
        }
        if (!opts.version) {
            opts.version = utilities.getVersion();
        }
        super(PodDisruptionBudget.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing PodDisruptionBudget resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new PodDisruptionBudget(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of PodDisruptionBudget.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === PodDisruptionBudget.__pulumiType;
    }
}
exports.PodDisruptionBudget = PodDisruptionBudget;
/** @internal */
PodDisruptionBudget.__pulumiType = 'kubernetes:policy/v1beta1:PodDisruptionBudget';
